### ISSUE TITLE
Provide actionable data to go with the "Some Agents within this Cluster are running an older version of Teleport" message

### ISSUE THREAD

**Issue by mdaniel** [gravitational/teleport#17488]
What would you like Teleport to do?

Show the information about the agents that it has identified are on older version instead of just saying there are some. A good middle ground would be to remind the user that `tctl inventory status --connected` will show information about the agents

What problem does this solve?

With a big enough fleet, it can be a lot of work to search everywhere to find the old agents that the Teleport cluster already knows exist, since that is the reason for the message

If a workaround exists, please include it.

Splunk? a lot of kubectl?

**Comment by zmb3** [comment]
`tctl inventory status --connected` will show you the versions of all connected agents.

**Comment by mdaniel** [comment]
this could be a separate bug, but tctl disagrees with my request:
```console
$ tctl inventory status --connected
ERROR: requires builtin admin role

$ tctl users ls | grep ^mdaniel
mdaniel@example.com       admin,dev
```
---

I'll update my request to compromise that the message could use the second part of that information message to say what you said instead of `Please update them`

**Comment by fspmarshall** [comment]
The `tctl inventory status --connected` command is currently hidden from `tctl help` because its a subset of an as of yet incomplete feature, so usage isn't very polished.

Currently, the command only works if you are running `tctl` locally on an auth server (I believe this is the reason for the confusing "builtin admin" error message).  The purpose of this command is the check which instances are specifically connected to the local auth server (i.e. you'd need to re-run it on all auth instances to see the complete set).

I'm working on adding support for listing the versions of all instances, regardless of which auth they or tctl is talking to.  Once this is in place we'll be stabilizing a collection of `tctl inventory <subcommand>` commands designed to let users view info about all teleport installations In the cluster regardless of the services they are runnings.

Will assign this issue to myself as part of that work.  The upgrade suggestion alert should point users to the proper `inventory` subcommand to let them see all their instances.  Possibly with some kind of filter so that users can immediately get the exact list of instances that are outdated (e.g. `tctl inventory ls --older-than=v1.2.3`).

**Comment by mdaniel** [comment]
> Currently, the command only works if you are running tctl locally on an auth server

We're on Teleport Cloud, so that is (to the best of my knowledge) never going to be an option for us

Do you want a separate ticket for the confusing error, or it's as you said: WIP and thus I wasn't _supposed_ to invoke it? :-D

**Comment by fspmarshall** [comment]
> If a workaround exists, please include it.

Until the `tctl inventory` family of commands stabilize, the best way to find outdated instances is to list each individual service spec (`node`, `app` etc) and check the `spec.version` field.  Something like this should work:

```
$ tctl get --format=json node | jq -r '.[] | .spec.hostname + " -> " + .spec.version'
foo.example.com -> 1.2.3
bar.example.com -> 2.3.4
...
```

**Comment by mdaniel** [comment]
I think I've mentioned this before in Slack, but for Cloud (or maybe kubernetes specifically? we have zero non-kubernetes instances in our fleet) that does nothing
```console
$ tctl get --format=json node
null
```

**Comment by fspmarshall** [comment]
> we have zero non-kubernetes instances

Ah, then try `tctl get --format=json kube_servers | jq -r '.[] | .spec.hostname + " -> " + .spec.version'`.  That should be the equivalent for your case.  Each service's state is currently tracked separately, which means different commands for different services.

> Do you want a separate ticket for the confusing error

That's alright, this should be something I fix as part of stabilizing the `inventory` commands.

> WIP and thus I wasn't supposed to invoke it? :-D

Haha, I wouldn't say not _supposed_ to, but hidden commands are usually only present because they are useful for us to be able to tell people to invoke during specific debugging scenarios, and might not be stable/polished enough to recommend to people for general use yet :)

**Comment by mdaniel** [comment]
I actually did have a sniff around before replying, and a lot of them have that same `null` output, which is why I suspected it was Cloud specific
```console
$ tctl get --format=json kube_servers
ERROR: unsupported resource: "kube_servers" - resources should be expressed as 'type/name', for example 'connector/github'
$ tctl get --format=json clusters
null
```

I was especially sad it was just `null` and not `{kind: List, spec: []}` or whatever as kubectl would have said for "I understood your query and the answer was  an empty list"

I didn't see anything that was the obvious equivalent to `kubectl api-resources` for `tctl` order to know what my choices were in that `get`. Maybe it's this list, buried in the example comment? https://goteleport.com/docs/access-controls/reference/?scope=cloud#roles That _seems_ to track because `tctl get db_server` emits something, and `kube_servers` doesn't show up in that comment

**Comment by fspmarshall** [comment]
@mdaniel Sorry about that!  I was working off of master.  Looks like you can't list kube heartbeats generically in v10.  Try `tctl kube ls` (just that, no `jq`).  That command should also show you the teleport version associated with each kube service (far right column).

**Comment by mdaniel** [comment]
For future generations who see this thread (a) `tctl kube ls --format=json` works, and (b) based on that output it seems the secret sauce was actually `tctl get --format=json kube_service` not `kube_servers` although the `kube_cluster` info is buried in the `.spec` so getting the cluster name back out requires some massaging

```console
$ tctl get --format json kube_service | \
  jq -r '[.[] | .spec.version+"\t"+.spec.kube_clusters[0].name] |sort |unique[]'
```
and that `|unique` business is required because there is one `kube_service` for every teleport-agent Pod, not Cluster (the `kube_service` `.metadata.name` is evidently the internal Teleport name for the agent because the uuid doesn't match up with any obvious part of the Pod)

**Comment by mighty-services** [comment]
Thanks for all the examples.  I am struggeling with this problem too, but not with databases or kubes, but with nodes. The only command working right now it the `tctl inventory status --connected`. It´s giving me good output, the only thing for me is - how do I find out/connect to the nodes, when I only see the id in the first column?

Is there a way to tweak this line to also show the nodename?

**Comment by stefankaufmann** [comment]
Are there any plans to add this information to the dashboard as well? The web-dashboard shows the warning, but does not show the version information of the clients. It feels like "I know there is a problem, but I won't tell you where it is ..."

**Comment by fspmarshall** [comment]
update: we are in the process of rolling out support for `tctl inventory ls --older-than=<version>` which will be the recommended way to do this doing forward. Ex:

```
$ tctl inventory ls --older-than=v13.2
ServerID                             Hostname          Services Version Upgrader 
------------------------------------ ----------------- -------- ------- -------- 
1e6578b6-9530-448e-8013-d32641324abb node1.example.com Node     v12.2.0 none     
7d2624d9-fb09-41d5-8dfd-d1f3d8112b9f node2.example.com Node     v12.3.0 unit     
...
```

 If you have a teleport cloud cluster, this feature should be enabled as of the latest release.  If you are managing your own control plane, you can enable the feature early by setting the env var `TELEPORT_UNSTABLE_ENABLE_INSTANCE_HB=yes` on all your auth servers (note: takes a couple minutes for all instances to show up after you do this).

Once we've ironed out some performance issues in very large clusters this will be enabled by default for everyone.  At that point, we'll be updating the alert to include the new recommended command.

**Comment by fspmarshall** [comment]
Closing due to https://github.com/gravitational/teleport/pull/29613 making this issue redundant.
